### ISSUE TITLE
ZigZagController with ColorEffect causes an exception

### ISSUE THREAD

**Issue by billy1380** [flame-engine/flame#3068]
### What happened?

When using the code:

```
  static final ColorEffect lockedBannerBlink = ColorEffect(
    opacityFrom: 1,
    opacityTo: 0,
    Colours.white,
    InfiniteEffectController(ZigzagEffectController(
      period: 1,
    )),
  );
```

the opacity is calculated at 1.066668

this causes the assertion to throw an exception in Color.withOpacity

### What do you expect?

I am not sure the ColorEffect should be generating values larger that 1 e.g. there is a max clamp to ensure that the value does not fall below 1/255 it might be worth having a min clamp to  ensure the value does not go above 1.

### How can we reproduce this?

```
  static final ColorEffect lockedBannerBlink = ColorEffect(
    opacityFrom: 1,
    opacityTo: 0,
    Colours.white,
    InfiniteEffectController(ZigzagEffectController(
      period: 1,
    )),
  );

await add(lockedBannerBlink);
```

### Execute in a terminal and put output into the code block below

Output of: flutter doctor -v
```
[✓] Flutter (Channel stable, 3.19.1, on macOS 14.2.1 23C71 darwin-x64, locale en-GB)
    • Flutter version 3.19.1 on channel stable at /Users/billy1380/git/flutter_stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision abb292a07e (2 weeks ago), 2024-02-20 14:35:05 -0800
    • Engine revision 04817c99c9
    • Dart version 3.3.0
    • DevTools version 2.31.1

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/billy1380/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/billy1380/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15E204a
    • CocoaPods version 1.14.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)

[✓] IntelliJ IDEA Ultimate Edition (version 2020.2.3)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] IntelliJ IDEA Community Edition (version 2023.3.4)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.87.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.84.0

[✓] Connected device (3 available)
    • Pixel 7 Pro (mobile) • 27211FDH3008HW • android-arm64  • Android 14 (API 34)
    • macOS (desktop)      • macos          • darwin-x64     • macOS 14.2.1 23C71 darwin-x64
    • Chrome (web)         • chrome         • web-javascript • Google Chrome 122.0.6261.112

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

### Affected platforms

All

### Are you interested in working on a PR for this?

- [ ] I want to work on this

**Comment by spydon** [comment]
Sounds like a good improvement to clamp at 1.0 opacity, would you like to provide a PR for this?

**Comment by ufrshubham** [comment]
This would be a nice improvement but @billy1380, since you are using a `ZigzagEffectController`, the `opacityFrom` value should be `0.5`.

**Comment by billy1380** [comment]
@spydon added unit test and fixed issue, pr here https://github.com/flame-engine/flame/pull/3069

**Comment by spydon** [comment]
Closed by #3069
